app/storage/metadata_parser.py

```diff
@@ -1,105 +1,117 @@
-        if self._validator:
-            # The parser methods throw exceptions on incorrect data
-            self._validator(original_value)
-
-    def generate(self):
-        if self._generator:
-            return self._generator()
-
-        return None
-
-
-metadata_fields = {
-    'user_id': MetadataField(),
-    'ru_ref': MetadataField(),
-    'ru_name': MetadataField(),
-    'eq_id': MetadataField(),
-    'collection_exercise_sid': MetadataField(),
-    'period_id': MetadataField(),
-    'period_str': MetadataField(),
-    'ref_p_start_date': MetadataField(validator=iso_8601_data_parser),
-    'ref_p_end_date': MetadataField(mandatory=False, validator=iso_8601_data_parser),
-    'form_type': MetadataField(),
-    'survey_url': MetadataField(mandatory=False),
-    'return_by': MetadataField(validator=iso_8601_data_parser),
-    'trad_as': MetadataField(mandatory=False),
-    'employment_date': MetadataField(mandatory=False, validator=iso_8601_data_parser),
-    'region_code': MetadataField(mandatory=False),
-    'language_code': MetadataField(mandatory=False),
-    'tx_id': MetadataField(mandatory=False, validator=uuid_4_parser, generator=id_generator),
-    'variant_flags': MetadataField(mandatory=False),
-    'roles': MetadataField(mandatory=False),
-    'case_id': MetadataField(mandatory=False),
-    'case_ref': MetadataField(mandatory=False),
-    'account_service_url': MetadataField(mandatory=False),
+VALIDATORS = {
+    'date': iso_8601_date_parser,
+    'uuid': uuid_4_parser,
+    'string': lambda *args: None,
+    'object': lambda *args: None
 }
 
+MANDATORY_METADATA = {
+    'eq_id': {
+        'validator': 'string'
+    },
+    'form_type': {
+        'validator': 'string'
+    },
+    'ru_ref': {
+        'validator': 'string'
+    },
+    'collection_exercise_sid': {
+        'validator': 'string'
+    },
+    'tx_id': {
+        'validator': 'uuid'
+    }
+}
 
-def parse_metadata(metadata_to_check):
-    parsed = {}
-    try:
-        for key, field in metadata_fields.items():
-            if key in metadata_to_check:
-                attr_value = metadata_to_check[key]
-                field.validate(attr_value)
-                logger.debug('parsing metadata', key=key, value=attr_value)
-            else:
-                logger.debug('generating metadata value', key=key)
-                attr_value = field.generate()
 
-            parsed[key] = attr_value
-    except (RuntimeError, ValueError, TypeError) as e:
-        logger.error('unable to parse metadata', exc_info=e)
-        raise InvalidTokenException('incorrect data in token')
+def parse_runner_claims(claims):
 
-    return clean_leading_trailing_spaces(parsed)
+    cleaned_claims = clean_leading_trailing_spaces(claims.copy())
+    cleaned_claims['tx_id'] = claims.get('tx_id', id_generator())  # Generate tx_id if not preset in claims
 
+    _validate_mandatory_schema_metadata(cleaned_claims, MANDATORY_METADATA)
+    metadata_validator(cleaned_claims)
```

Comment:
If you just passed `MANDATORY_METADATA` here you wouldn't need line 81 (`validators = validators or MANDATORY_METADATA`) and you can remove the None default.

Then these two lines are the same as the whole of method  `validate_schema_metadata`. It might therefore be worth considering just calling a generic `validate_metadata` and just call it twice, once for each of the two types of metadata that you're validating. 

If you do that it might also be worth considering renaming method `metadata_validator` to `_validate_metadata_values`

```
def validate_metadata(claims, required_metadata):
    _validate_mandatory_metadata(claims, required_metadata)
    _validate_metadata_values(claims, required_metadata)
```

Reply:
👍